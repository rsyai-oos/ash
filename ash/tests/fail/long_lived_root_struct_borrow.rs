use ash::vk::{self, TaggedStructure as _};

fn main() {
    // src/vk.rs holds a positive version of this test, demonstrating that `api` can be adequately
    // accessed when `pdev_props` is no longer live.

    let mut layers = vec![];
    let mut api =
        vk::PhysicalDeviceLayeredApiPropertiesListKHR::default().layered_apis(&mut layers);
    let pdev_props = vk::PhysicalDeviceProperties2::default().push(&mut api);

    // Access to either variable is disallowed because both are mutably borrowed in pdev_props
    dbg!(&api);
    dbg!(&layers);

    dbg!(pdev_props); // Holds a borrow on api and layers
}
