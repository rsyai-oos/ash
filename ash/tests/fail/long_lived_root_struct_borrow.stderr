error[E0502]: cannot borrow `api` as immutable because it is also borrowed as mutable
  --> tests/fail/long_lived_root_struct_borrow.rs:13:10
   |
10 |     let pdev_props = vk::PhysicalDeviceProperties2::default().push(&mut api);
   |                                                                    -------- mutable borrow occurs here
...
13 |     dbg!(&api);
   |          ^^^^ immutable borrow occurs here
...
16 |     dbg!(pdev_props); // Holds a borrow on api and layers
   |          ---------- mutable borrow later used here

error[E0502]: cannot borrow `layers` as immutable because it is also borrowed as mutable
  --> tests/fail/long_lived_root_struct_borrow.rs:14:10
   |
 9 |         vk::PhysicalDeviceLayeredApiPropertiesListKHR::default().layered_apis(&mut layers);
   |                                                                               ----------- mutable borrow occurs here
...
14 |     dbg!(&layers);
   |          ^^^^^^^ immutable borrow occurs here
15 |
16 |     dbg!(pdev_props); // Holds a borrow on api and layers
   |          ---------- mutable borrow later used here
